py-rse/package-py.Rmd

```diff
@@ -20,445 +20,110 @@ This lesson shows you how to use Python's tools to create and share libraries of
-    ...as before...
-
-if __name__ == '__main__':
-    if len(sys.argv) == 1:
-        print(USAGE)
-    else:
-        values = [int(a) for a in sys.argv[1:]]
-        result = is_zipf(values)
-        print('{}: {}'.format(result, values))
-    sys.exit(0)
-```
-
-Here,
-the code guarded by `if __name__ == '__main__'` isn't executed when the file loaded by something else.
-We can test this by re-running `use.py` as before:
-the usage message doesn't appear,
-which means the main block wasn't executed,
-which is what we want.
-
-### Exercise: What is `__name__` in different contexts?
-
-1. Start an interactive Python interpreter, and run `print(__name__)`.
-What do you get?
-2. Add the statement `print(__name__)` to the end of the file `zipf.py`
-and from the Python interpreter, run `import zipf`.  What is the name
-printed?
-
-### Exercise: Turn `use.py` into a unit test
-
-1. Turn the `use.py` script into a `pytest` unit test; name the file `test_zipf.py`.
-
-## How do I create a package? {#py-rse-package-py-creating}
-
-Let's say we wanted to add some capabilities to what
-we've built: we want to be able to generate noisy
-Zipf distributions—noise within the same default `RELATIVE_ERROR`—and
-we want to add a Zipf-checker that can handle unsorted inputs:
-
-```python
-...as before...
-
-def make_zipf(length):
-    assert length > 0, 'Zipf distribution must have at least one element'
-    result = [1/(1 + i) for i in range(length)]
-    return result
-
-def make_noisy_zipf(length, rel=RELATIVE_ERROR):
-    data = make_zipf(length)
-    minnoise = 1.0 - rel/2
-    maxnoise = 1.0 + rel/2
-    for i in range(length):
-        data[i] = data[i] * random_uniform(minnoise, maxnoise)
-    return data
-
-def is_zipf(hist, rel=RELATIVE_ERROR):
-    assert len(hist) > 0, 'Cannot test Zipfiness without data'
-    scaled = [h/hist[0] for h in hist]
-    perfect = make_zipf(len(hist))
-    return scaled == approx(perfect, rel=rel)
-
-def is_unsorted_zipf(hist, rel=RELATIVE_ERROR):
-    sortedhist = sorted(hist)
-    return is_zipf(unsorted, rel)
-
-...as before...
-```
-
-As our package grows,
-we should split its source code into multiple files,
-and this is now starting to look like two broad types of
-related functionality: generating a Zipf distribution, and
-checking against a Zipf distribution.  But they are related;
-and what's more, there are constants that would be common
-across the two pieces - the `RELATIVE_ERROR` constant.
-
-If we want multiple files, but they are related and
-and have some common items, that's a good indication that
-we are outgrowing a single-file module and we want a Python
-package.
-
-A package in Python is no more than a directory containing
-Python code, and by convention a Python file with a special
-name, `__init__.py` (this was a requirement before Python 3.3).
+```text
+zipfs_law
+├── zipfs_law
+│   ├── __init__.py
+│   ├── collate.py
+│   ├── countwords.py
+│   ├── mymodule.py
```

Comment:
It occurs to me that if we are packaging `zipfs_law` for release then we probably want a better / more realistic name for `mymodule.py`. Perhaps `csv_tools.py`?

Reply:
Opened #341